### ISSUE TITLE
HTTPS for local development

### ISSUE THREAD

**Issue by onnimonni** [IFTTT/dash#25]
(Moved from #24)
I hate that so many use https in production but they don't in development. This causes multiple errors constantly. For example devs don't notice mixed content errors.

We could selfsign a wildcard certificate with something like `*.dash.test` or `*.d.test` and then use it for `jwilder/nginx-proxy`. Then if we use domains like `my-application.dash.test` for containers they would have https enabled by default.
[jwilder/nginx-proxy already supports this](https://github.com/jwilder/nginx-proxy#sni), we only need to mount the generated cert folder into the container. These certificates could be stored locally in `~/.dash/` for example. If stored locally we could add more domains later with a command something like `$ dev ssl add *.site.test`.

This would also allow testing the real certificates by changing `production.address` in /etc/hosts.

I checked and tested but it's not possible to use self-signed certificate for tld eg `*.test`. Safari will accept this without a question (if it's in keychain) but chrome is more clever and denies this.

You can trust self-generated keys in OS-X by:
```ruby
system "sudo security add-trusted-cert -d -r trustRoot -k '/Library/Keychains/System.keychain' #{ssl_cert_path}/development.crt"
```

I can do PR for this in couple of days. I just want to hear your opinions about these as well.

**Comment by benjamin-smith** [comment]
I think it makes sense to not use the machine cert, but rather ask users to create their own, or have the bootstrap script handle this task. I'd rather not reuse an existing cert for convenience, as that may cause unexpected issues of some kind. Separating the certs will also give us more control over the Dash cert, rather than rely on what machine gives us.

**Comment by onnimonni** [comment]
Or we could just parse the docker-compose.yml and create matching certs for the domain names in nginx-proxy mounted certs folder.

**Comment by onnimonni** [comment]
I mean we could do this in: $ dev up

**Comment by silvamerica** [comment]
Yeah, I'll backpeddle on using the machine cert, as all of your points against it are good ones. 

@onnimonni would we be able to have matching certs in `nginx-proxy` without scanning all possible `docker-compose.yml` files? I'm not exactly sure what you're suggesting. Maybe the `nginx-proxy` module could create certs on the fly with a trusted CA that we create during bootstrapping?

**Comment by onnimonni** [comment]
I think I could make a new container like this: [JrCs/docker-letsencrypt-nginx-proxy-companion](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion) but only for local development and self-signed certs.

The container can listen `docker.sock` and could generate new certificates on the fly and our install script should only add the self-signed ca to trusted list. Then pass the ca for this container. And mount the same `/certs/` folder inside this container and jwilder/nginx-proxy.

Should these certificates be stored in project directory or in user directory `~/.dash`?

**Comment by silvamerica** [comment]
`~/.dash` hasn't been something that we've used before, but I'm not opposed to it. For that matter, I mostly arbitrarily picked `/usr/local/dev-env`. Does it make more sense to put them in the project directory, but move the project directory to ~? (We could make another topic for that)

**Comment by onnimonni** [comment]
I made this certificate generator container during weekend: https://github.com/onnimonni/signaler

If you add `HTTPS_HOST: example.dev` into `docker-compose.yml` this container will create `example.dev` certificate automatically.

I'm doing the HTTPS already in our fork:
https://github.com/devgeniem/gdev/blob/master/docker/services.yml

I'm using user home folder for certs and mounting the folder in nginx and signaler containers.

I was in such a hurry so I didn't make a pull request (at least not yet), would something similiar work in dash?
